src/Helpers/TextHelper.php

```diff
@@ -21,4 +26,271 @@ public static function estimateTokenCount(string $text): int {
+   *
+   * @return int
+   *   Estimated number of characters in the text.
+   */
+  public static function estimateStringLength(int $token_count): int {
+    return (int) floor(.9 * 4 * $token_count);
+  }
+
+  /**
+   * Split a text in paragraphs.
+   *
+   * @param string $text
+   *   Text.
+   *
+   * @return array
+   *   Array of one or more strings.
+   */
+  public static function splitTextInParagraphs(string $text) : array {
+    $text = trim($text);
+    if (empty($text)) {
+      return [];
+    }
+
+    return preg_split('/\n{2,}/u', $text, -1, \PREG_SPLIT_NO_EMPTY);
+  }
+
+  /**
+   * Split paragraph in sentences.
+   *
+   * @param string $text
+   *   Text.
+   * @param string $pattern_id
+   *   Pattern to use.
+   *
+   * @return array
+   *   Array of one or more strings with meta data.
+   */
+  public static function splitParagraphsInSentences(string $text, string $pattern_id = 'default') : array {
+    $text = trim($text);
+    if (empty($text)) {
+      return [];
+    }
+
+    if ($pattern_id == 'default') {
+      $pattern_id = 'punctuation';
+    }
+
+    // Possible patterns to use.
+    $patterns = [
+      'punctuation' => '/([;.!?。؟]+)\s+/u',
+      'split_capital' => '~[.?!]+\K\s+(?=[A-Z])~',
+    ];
+    $pattern = $patterns[$pattern_id] ?? $patterns['punctuation'];
+
+    $text = preg_replace('/\s+/u', ' ', $text);
+    $text = trim(preg_replace($pattern, "$1\n", $text));
+
+    $sentences = [];
+    foreach (preg_split('/\n+/u', $text, -1, \PREG_SPLIT_NO_EMPTY) as $line) {
+      $sentences[] = [
+        'text' => $line,
+        'token_count' => self::estimateTokenCount($line),
+        'char_count' => mb_strlen($line),
+      ];
+    }
+
+    return $sentences;
+  }
+
+  /**
+   * Split a text based in lines of <length> characters.
+   *
+   * @param string $text
+   *   Text.
+   * @param int $length
+   *   Max lenght.
+   * @param string $unit
+   *   Unit for length.
+   * @param string $pattern_id
+   *   Pattern to use.
+   *
+   * @return array
+   *   Array of one or more strings.
+   */
+  public static function splitInLines(string $text, int $length, string $unit = self::UNIT_TOKEN, string $pattern_id = 'default') : array {
+    $output = [];
+    $text = trim($text);
+
+    if (empty($text)) {
+      return $output;
+    }
+
+    if ($length <= 0) {
+      return $output;
+    }
+
+    $paragraphs = self::splitTextInParagraphs($text);
+    if (empty($paragraphs)) {
+      return $output;
+    }
+
+    $str_length = $length;
+    if ($unit == self::UNIT_TOKEN) {
+      $str_length = self::estimateStringLength($length);
+    }
+
+    foreach ($paragraphs as $paragraph) {
+      $sentences = self::splitParagraphsInSentences($paragraph, $pattern_id);
+
+      foreach ($sentences as $sentence) {
+        if (self::getLength($sentence, $unit) < $length) {
+          $output[] = $sentence;
+          continue;
+        }
+
+        // Line is too long.
+        while (mb_strlen($sentence) > 0) {
```

Comment:
`$sentence` is an array (see `:: splitParagraphsInSentences()`) so this throws a TypeError exception.

It's missing something like `$sentence = $sentence['text'];` before the `while`.

Reply:
Code rewritten